### PR TITLE
@log_errors for async tasks

### LINKED ISSUE #8292: `@log_errors` can't decorate asyncio tasks
This logs the exception:
```python
@log_errors
async def f():
    assert False

await f()
```
but this doesn't:
```python
await asyncio.create_task(f())
```

From debugging, I see that here:
https://github.com/dask/distributed/blob/095a09d4e71385f9ceed45de0780480dd26b9015/distributed/utils.py#L831-L836

`modname` is `asyncio.events` instead of the module that raised the error.

### PR DESCRIPTION
**Pull request by crusaderky**
- Closes #8292

### CODE REVIEW

File: distributed/utils.py [hunk 1]
```diff
@@ -783,7 +784,7 @@ def func(...):
         Set to True to break into the debugger in case of exception
     unroll_stack: int, optional
         Number of levels of stack to unroll when determining the module's name for the
-        purpose of logging. Normally you should omit this. Set to 2 if you are writing a
+        purpose of logging. Normally you should omit this. Set to 1 if you are writing a
```
Comment (crusaderky) [inline distributed/utils.py:787]: log_errors is not linked anywhere in sphinx, so I believe it should be treated as internal.

File: distributed/utils.py [hunk 2]
```diff
@@ -783,13 +784,33 @@ def func(...):
...
+def _fname_to_modname(fname: str) -> str:
+    try:
+        return _fname_to_modname_map[fname]
+    except KeyError:
+        pass
+
+    for modname, mod in sys.modules.items():
+        fname2 = getattr(mod, "__file__", None)
+        if fname2:
+            _fname_to_modname_map[fname2] = modname
```
Comment (crusaderky) [inline distributed/utils.py:806]: This isn't great but it's called only on unexpected errors

File: distributed/utils.py [hunk 3]
```diff
@@ -820,16 +842,15 @@ def wrapper(*args, **kwargs):
...
 
-    def __exit__(self, exc_type, exc_value, traceback):
+    def __exit__(self, exc_type, exc_value, tb):
         from distributed.comm import CommClosedError
 
         if not exc_type or issubclass(exc_type, (CommClosedError, gen.Return)):
             return
 
-        stack = inspect.stack()
+        stack = traceback.extract_tb(tb)
```
Comment (crusaderky) [inline distributed/utils.py:851]: Unsure why this gives the correct result whereas this other one doesn't:

```python
frame = tb.tb_frame
for _ in range(self.unroll_stack):
    frame = frame.f_prev
mod = inspect.getmodule(frame)
modname = mod.__name__
```

File: distributed/utils.py [hunk 4]
```diff
@@ -820,16 +842,15 @@ def wrapper(*args, **kwargs):
...
         frame = stack[self.unroll_stack]
-        mod = inspect.getmodule(frame[0])
-        modname = mod.__name__
+        modname = _getmodulename_with_path(frame.filename)
 
         try:
             logger = logging.getLogger(modname)
```
Comment (crusaderky) [inline distributed/utils.py:856]: This is not ideal e.g. for the shuffle internal modules. We can improve it but it's the same as how main works so it should be treated as out of scope for this PR.

**Review by hendrikmakait — APPROVED** [review]

**Pull request merged**